### PR TITLE
[DataGrid] Default toolbar setup

### LINKED ISSUE #516: [DataGrid] Add toolbar
## Summary 💡

Introduce a toolbar to perform complementary actions. 

## Examples 🌈

Items the toolbar can include:

- Filtering
- Grouping
- Sorting
- Column order
- Column visibility
- Display density
- Export/import records
- Quick search
- Add new record

## Motivation 🔦

The data grid is meant to display records, many operations & manipulator can be performed on these records, a toolbar is a great place to manage it. The Material Design specification quickly [mentions it](https://material.io/components/data-tables#anatomy) but doesn't provide guidance that we can truly leverage.

We could leverage the composability of React. We could imagine introducing new components, like a `<DataGridToolar>` component that would take the `state` object in. For instance, Stripe and Gmail introduce tabs between the toolbar and the content of the grid.

## Benchmark

- https://airtable.com, active items are colored. It works great.

Capture d’écran 2020-10-31 à 16 25 44

- https://ag-grid.com, side panel, the UX is horrible, to ignore.

Capture d’écran 2020-10-31 à 16 26 30

- https://try.firetable.io/table/mvpResources, works OK-ish

Capture d’écran 2020-10-31 à 16 27 59

- https://elastic.github.io/eui/#/tabular-content/data-grid, works well

Capture d’écran 2020-10-31 à 16 29 32

- https://baseweb.design/components/unstable-data-table#data-table, displaying the filters directly, doesn't scale, not so great.

Capture d’écran 2020-10-31 à 16 37 58

- https://ads.google.com/

Capture d’écran 2020-10-31 à 16 54 53

### PR DESCRIPTION
**Pull request by DanailH**
This pull request aims to add an empty toolbar that will sit at the top of the data grid (above the columns and the header component it such is provided). It is structured in a way that will allow us to be easily extracted in the future if we decided to do so. To add buttons (features) in the `Toolbar` on needs to create a separate component for that feature (eg. `ColumnPicker`) and implement the functionality there. After it is done the following structure should be followed when adding the feature to the toolbar:

```javascript
<DefaultToolbar>
  <ColumnPicker />
  ...
</DefaultToolbar>
```

Questions that need to be answered in this Draft PR:
- What to do if there are no buttons/features added (like it is now). The toolbar is visible although it is empty. I think it would be better to hide it if no children are provided.
- Once we agree that it is a viable solution I will add the needed tests and documentation, for now, there is none.
- The styling of the Toolbar - for now I just added a fixed `min-height`, centered the items vertically, added a border, and a background. The `min-height` matches the one we have for an MUI `button` with size `small` (a similar approach was taken for the footer where the `min-height` matches the height of the `TablePagination` component).

Closes #516

### CODE REVIEW

**Review by oliviertassinari — COMMENTED** [review]
Looks like a good skeleton. Note that we need to fix https://github.com/mui-org/material-ui-x/issues/502 before we can display anything in the header. I believe the solution is to reorganize the grid to have the AutoSizer nested

**Comment by DanailH** [comment]
Nice, ok let's see what @dtassone will say about the skeleton because I need this merged in order to start with either the column picker or the density selector. In the meantime, I can take a look at the `AutoSizer` issue.

**Comment by DanailH** [comment]
If this will be the way to go should I proceed with adding the documentation, hiding the toolbar if it is empty, and convert it to a normal PR?

File: packages/grid/_modules_/grid/GridComponent.tsx [hunk 1]
```diff
@@ -119,6 +120,11 @@ export const GridComponent = React.forwardRef<HTMLDivElement, GridComponentProps
             >
               <ApiContext.Provider value={apiRef}>
                 <OptionsContext.Provider value={gridState.options}>
+                  {customComponents.toolbarComponent || (
```
Comment (dtassone) [inline packages/grid/_modules_/grid/GridComponent.tsx:123]: you can do the check for hideToolbar here

File: packages/grid/_modules_/grid/components/DefaultToolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import * as React from 'react';
+import { useGridSelector } from '../hooks/features/core/useGridSelector';
+import { optionsSelector } from '../hooks/utils/useOptionsProp';
+import { ApiContext } from './api-context';
+import { GridToolbar } from './styled-wrappers/GridToolbar';
+
+// TODO: We can list separate features here, eg. column picker
+export interface DefaultToolbarProps {}
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/DefaultToolbar.tsx:8]: Not sure I understand the purpose of this interface

File: packages/grid/_modules_/grid/components/DefaultToolbar.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    const { children } = props; // Extranct the child components from the props
+    const apiRef = React.useContext(ApiContext);
+
+    // TODO: This part won't be needed if the component is separate
+    const options = useGridSelector(apiRef, optionsSelector);
+    if (options.hideToolbar) {
+      return null;
+    }
+
+    return <GridToolbar ref={ref}>{children}</GridToolbar>;
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/DefaultToolbar.tsx:21]: Why do we need a defaultToolbar and a GridToolbar?

File: packages/grid/_modules_/grid/GridComponent.tsx (hunk 1 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/GridComponent.tsx]: Done

File: packages/grid/_modules_/grid/components/DefaultToolbar.tsx (hunk 2 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/DefaultToolbar.tsx]: My initial idea was to list the separate components here and then add logic if there are provided from outside to overwrite the ones we provide as default but now I don't think it is needed since developers can provide an entire custom Toolbar

File: packages/grid/_modules_/grid/components/DefaultToolbar.tsx (hunk 3 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/DefaultToolbar.tsx]: No need for it anymore, I followed the logic for the footer but since we don't have any additional functionality i removed the `DefaultToolbar` and used the `GridToolbar` directly.

File: docs/src/pages/components/data-grid/rendering/rendering.md [hunk 4]
```diff
@@ -273,6 +274,17 @@ This demo replaces it with the [Pagination](/components/pagination/) component.
 
 {{"demo": "pages/components/data-grid/rendering/CustomPaginationGrid.js", "bg": "inline"}}
 
+### Toolbar
+
+The grid exposes props to hide specific elements of the UI:
+
+- `hideToolbar`: If `true`, the toolbar component is hidden.
+
+By default, we provide a default toolbar with opinianted functionality. You can replace our default toolbar by providing your own implementation.
```
Comment (dtassone) [inline docs/src/pages/components/data-grid/rendering/rendering.md:283]: opinianted??

**Review by dtassone — APPROVED** [review]

File: docs/src/pages/components/data-grid/rendering/rendering.md (hunk 4 above)
Comment (DanailH) [reply docs/src/pages/components/data-grid/rendering/rendering.md:283]: I didn't know how to call it. Maybe that is not the best way to describe it, but we will put functionality in it that we think is useful. Maybe just having it as `By default, we provide a default toolbar.` is better.

Comment (DanailH) [reply docs/src/pages/components/data-grid/rendering/rendering.md]: done

File: docs/src/pages/components/data-grid/rendering/rendering.md [hunk 5]
```diff
@@ -245,6 +245,7 @@ As part of the customization API, the grid allows you to replace and override ne
 - `noRowsOverlay`: The component rendered when the rows react prop is empty or [].
 - `footer`: The component rendered below the viewport.
 - `pagination`: The component rendered for the pagination feature.
+- `toolbar`: The component rendered above the column header bar and the custom `header` component if provided.
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/rendering/rendering.md:248]: They are sorted by vertical display order. I assume it should be before `header`

File: docs/src/pages/components/data-grid/rendering/rendering.md [hunk 6]
```diff
@@ -273,6 +274,17 @@ This demo replaces it with the [Pagination](/components/pagination/) component.
 
 {{"demo": "pages/components/data-grid/rendering/CustomPaginationGrid.js", "bg": "inline"}}
 
+### Toolbar
+
+The grid exposes props to hide specific elements of the UI:
+
+- `hideToolbar`: If `true`, the toolbar component is hidden.
+
+By default, we provide a default toolbar. You can replace it by providing your own implementation.
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/rendering/rendering.md:283]:
```suggestion
You can replace the default toolbar by providing your own implementation.
```

File: packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts [hunk 7]
```diff
@@ -250,6 +250,14 @@ export const useStyles = makeStyles(
           borderRadius: theme.shape.borderRadius,
           opacity: theme.palette.action.disabledOpacity,
         },
+        '& .MuiDataGrid-toolbar': {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts:253]: It would be better to keep the CSS organized by logical order.

In this case, I think that we can move the style to the toolbar, the toolbar doesn't & shouldn't rerender at 60 FPS. But that could be an effort for another time.

File: packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+  if (!children) {
+    return null;
+  }
+
+  return (
+    <div ref={ref} className={classnames('MuiDataGrid-toolbar', className)} {...other}>
+      {children}
+    </div>
+  );
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx:20]: This behavior would break the standard rendering logic of the core components. Why render null if no children are provided?

```suggestion
  const { className, ...other } = props;

  return (<div ref={ref} className={classnames('MuiDataGrid-toolbar', className)} {...other} />);
```

**Review by oliviertassinari — COMMENTED** [review]
Why would somebody use the `toolbar` over the `header`?

File: packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx (hunk 8 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx:20]: > Why would somebody use the `toolbar` over the `header`?

Idk, I can move it under the header. Now that I think about it it does sound logical to be under it.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx:20]: Yeah, sounds better, no need for it;

Comment (DanailH) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx:20]: > This behavior would break the standard rendering logic of the core components. Why render null if no children are provided?

This is a temporary logic. It should be removed once there at least 1 thing in the toolbar. If I remove this check now the toolbar will render and it will be empty, It will only take space. I did it that way so that if we want to release next Tuesday lets say and the `ColumnPicker` or `DencitySelector` or the filtering are not yet ready or merged we can still do it without introducing an empty toolbar that is aways visible by default.

File: packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts (hunk 7 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts:253]: ok so I can move this up in the file

File: packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx (hunk 8 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx:20]: > This is a temporary logic. It should be removed once there at least 1 thing in the toolbar.

The more I think about it, the more I think this is the wrong direction, even for the short term. Developers will want to decide which action in the toolbar they display. How will we do that? With a boolean for each feature? A composable API sounds simpler

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx:20]: Check this one: https://github.com/mui-org/material-ui-x/issues/575.

Comment (DanailH) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridToolbar.tsx:20]: I agree the better approach would have been to have it as a separate component wrapped around a provider but we chose to do it like this. Also from our discussions, I didn't expect to have the toolbar inside the Grid for a long time but more of a temporary solution to speed up the feature development.

But yes you are right it is bad to have a boolean per feature to disable them. For the current implementation I was more thinking in terms of - if you don't want certain features, provide your own toolbar because otherwise, the public API will be just too big and badly designed.

File: packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts (hunk 7 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts]: done

File: docs/src/pages/components/data-grid/rendering/rendering.md (hunk 5 above)
Comment (DanailH) [reply docs/src/pages/components/data-grid/rendering/rendering.md]: done

File: docs/src/pages/components/data-grid/rendering/rendering.md [hunk 9]
```diff
@@ -241,6 +241,7 @@ You can disable column virtualization by setting the column buffer to a higher n
 As part of the customization API, the grid allows you to replace and override nested components with the `components` prop. The prop accepts an object with the following keys:
 
 - `header`: The component rendered above the column header bar.
+- `toolbar`: The component rendered above the column header bar and the custom `header` component if provided.
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/rendering/rendering.md:244]:
```suggestion
```

File: packages/grid/_modules_/grid/hooks/features/useComponents.tsx [hunk 10]
```diff
@@ -83,6 +92,7 @@ export const useComponents = (
 
   return {
     headerComponent,
+    toolbarComponent,
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/useComponents.tsx:95]:
```suggestion
```

File: packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts [hunk 11]
```diff
@@ -41,6 +41,7 @@ export function useOptionsProp(apiRef: ApiRef, props: GridComponentProps): GridO
       disableExtendRowFullWidth: props.disableExtendRowFullWidth,
       headerHeight: props.headerHeight,
       hideFooter: props.hideFooter,
+      hideToolbar: props.hideToolbar,
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts:44]:
```suggestion
```

File: packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts [hunk 12]
```diff
@@ -86,6 +87,7 @@ export function useOptionsProp(apiRef: ApiRef, props: GridComponentProps): GridO
       props.disableExtendRowFullWidth,
       props.headerHeight,
       props.hideFooter,
+      props.hideToolbar,
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts:90]:
```suggestion
```

File: packages/grid/_modules_/grid/models/gridComponentOverridesProp.tsx [hunk 13]
```diff
@@ -17,6 +17,10 @@ export interface GridComponentOverridesProp {
    * No rows overlay component rendered when the grid has no rows.
    */
   noRowsOverlay?: React.ElementType<ComponentProps>;
+  /**
+   * Toolbar component rendered at the top of the grid viewport.
+   */
+  toolbar?: React.ElementType<ComponentProps>;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridComponentOverridesProp.tsx:23]:
```suggestion
```

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 14]
```diff
@@ -147,6 +147,11 @@ export interface GridOptions {
    * @default false
    */
   hideFooter?: boolean;
+  /**
+   * If `true`, the toolbar component is hidden.
+   * @default false
+   */
+  hideToolbar?: boolean;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:154]:
```suggestion
```

File: packages/storybook/src/stories/playground/customize-components.stories.tsx [hunk 15]
```diff
@@ -317,3 +317,24 @@ export function StyledColumns() {
...
+  const data = useData(2000, 200);
+
+  return (
+    <div className="grid-container">
+      <XGrid
+        rows={data.rows}
+        columns={data.columns}
+        pageSize={33}
+        components={{
+          toolbar: ToolbarComponent,
```
Comment (oliviertassinari) [inline packages/storybook/src/stories/playground/customize-components.stories.tsx:335]:
```suggestion
          header: ToolbarComponent,
```

File: docs/src/pages/components/data-grid/rendering/CustomToolbarGrid.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    dataSet: 'Commodity',
+    rowLength: 100,
+    maxColumns: 6,
+  });
+
+  return (
+    <div style={{ height: 400, width: '100%' }}>
+      <DataGrid
+        components={{
+          toolbar: CustomToolbar,
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/rendering/CustomToolbarGrid.js:20]:
```suggestion
          header: CustomToolbar,
```

**Comment by oliviertassinari** [comment]
I have updated the pull-request to go into the header direction:

```jsx
<DataGrid
  components={{
    header:{
     <GridToolbar>
       <GridToolbarFilter />
       <GridToolbarDensity />
    </GridToolbar>
    }
  }}
/>
```

**Comment by DanailH** [comment]
> I have updated the pull-request to go into the header direction:
> 
> ```js-jsx
> <DataGrid
>   components={{
>     header: () => (
>       <GridToolbar>
>         <GridToolbarFilter />
>         <GridToolbarDensity />
>       </GridToolbar>
>     )
>   }}
> />
> ```

Ok so probably I misunderstood your previous comment. You want basically for the toolbar to be the header in terms of the API we expose.

**Comment by DanailH** [comment]
Ok, I've cleaned it up further, so now to overwrite the `GridToolbar` one needs to provide a `header` component.

File: docs/src/pages/components/data-grid/rendering/rendering.md [hunk 17]
```diff
@@ -273,6 +273,13 @@ This demo replaces it with the [Pagination](/components/pagination/) component.
 
 {{"demo": "pages/components/data-grid/rendering/CustomPaginationGrid.js", "bg": "inline"}}
 
+### Toolbar
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/rendering/rendering.md:276]:
```suggestion
### Header
```

File: docs/src/pages/components/data-grid/rendering/rendering.md [hunk 18]
```diff
@@ -273,6 +273,13 @@ This demo replaces it with the [Pagination](/components/pagination/) component.
...
+
+You can replace the default toolbar by providing a `header` component.
+This demo showcases how this can be achieve.
+
+{{"demo": "pages/components/data-grid/rendering/CustomToolbarGrid.js", "bg": "inline"}}
+
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/rendering/rendering.md:282]: I would suggest we document the components in the vertical order they are rendered in the grid, so moving this section before `### Loading overlay`.

**Review by oliviertassinari — COMMENTED** [review]
Great, I had a sync with @dtassone, and we are all aligned for pushing further with this API :)

File: docs/src/pages/components/data-grid/rendering/rendering.md (hunk 18 above)
Comment (DanailH) [reply docs/src/pages/components/data-grid/rendering/rendering.md]: Done

**Comment by DanailH** [comment]
> Great, I had a sync with @dtassone, and we are all aligned for pushing further with this API :)

Nice, that is still easily abstractable tho, so if we decide to go with the `Provider` composition it will still be fairly easy to do it.

**Review by oliviertassinari — APPROVED** [review]

**Comment by max4211** [comment]
Hello! I am attempting to use the GridToolbar customization to include both the filters, density, and a custom button for export. Wondering if you have any pointers? Right now the button is not being created.

`        components={{
          Header: () => (
            <GridToolbar>
              <Button onClick={handleExport}>Export</Button>
            </GridToolbar>
          ),
        }}`

File: docs/src/pages/components/data-grid/rendering/rendering.md [hunk 19]
```diff
@@ -275,11 +275,7 @@ This demo replaces it with the [Pagination](/components/pagination/) component.
 
 ### Toolbar
 
-The grid exposes props to hide specific elements of the UI:
-
-- `hideToolbar`: If `true`, the toolbar component is hidden.
-
-You can replace the default toolbar by providing your own implementation.
+You can replace the default toolbar by providing a `header` component.
```
Comment (dtassone) [inline docs/src/pages/components/data-grid/rendering/rendering.md]: I'm lost with this. I thought we would add a toolbar slot

File: packages/grid/_modules_/grid/hooks/features/useComponents.tsx [hunk 20]
```diff
@@ -39,14 +39,6 @@ export const useComponents = (
     [componentOverrides, componentProps],
   );
 
-  const toolbarComponent = React.useMemo(
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/useComponents.tsx]: Why?

File: docs/src/pages/components/data-grid/rendering/rendering.md (hunk 19 above)
Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/rendering/rendering.md]: This PR is 3 months old
